File: scripts/status.py
#!/usr/bin/env python3
"""Print a user-oriented summary of the Kaimahi resources in one cluster."""

import argparse
import json
import subprocess
import sys


def kubectl(command, *args, required=True):
    proc = subprocess.run(
        [*command, *args, "-o", "json"],
        check=False,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    if proc.returncode != 0:
        if required:
            print(proc.stderr.rstrip(), file=sys.stderr)
            raise SystemExit(proc.returncode)
        return []
    return json.loads(proc.stdout).get("items", [])


def condition(resource, name):
    conditions = resource.get("status", {}).get("conditions", [])
    match = next((item for item in conditions if item.get("type") == name), None)
    if match is None:
        return "-"
    return "yes" if match.get("status") == "True" else "no"


def table(headers, rows):
    widths = [len(header) for header in headers]
    for row in rows:
        for index, value in enumerate(row):
            widths[index] = max(widths[index], len(value))
    print("  " + "  ".join(value.ljust(widths[index]) for index, value in enumerate(headers)))
    for row in rows:
        print("  " + "  ".join(value.ljust(widths[index]) for index, value in enumerate(row)))


def pod_summary(pods):
    ready = 0
    restarts = 0
    rows = []
    for pod in sorted(pods, key=lambda item: item["metadata"]["name"]):
        statuses = pod.get("status", {}).get("containerStatuses", [])
        is_ready = bool(statuses) and all(item.get("ready", False) for item in statuses)
        pod_restarts = sum(item.get("restartCount", 0) for item in statuses)
        ready += int(is_ready)
        restarts += pod_restarts
        rows.append(
            [
                pod["metadata"]["name"],
                "yes" if is_ready else "no",
                pod.get("status", {}).get("phase", "Unknown"),
                str(pod_restarts),
            ]
        )
    return ready, restarts, rows


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--context", required=True)
    parser.add_argument("--target", required=True)
    args = parser.parse_args()
    command = ["kubectl", "--context", args.context]

    agents = kubectl(command, "-n", "kagent", "get", "agents")
    models = kubectl(command, "-n", "kagent", "get", "modelconfigs")
    kagent_pods = kubectl(command, "-n", "kagent", "get", "pods")
    ollama_pods = kubectl(command, "-n", "ollama", "get", "pods", required=False)
    plane_pods = kubectl(command, "-n", "kaimahi", "get", "pods", required=False)

    agent_rows = []
    for agent in sorted(agents, key=lambda item: item["metadata"]["name"]):
        declarative = agent.get("spec", {}).get("declarative", {})
        servers = []
        for tool in declarative.get("tools") or []:
            name = (tool.get("mcpServer") or {}).get("name")
            if name:
                servers.append(name)
        agent_rows.append(
            [
                agent["metadata"]["name"],
                condition(agent, "Ready"),
                condition(agent, "Accepted"),
                declarative.get("modelConfig", "-"),
                ",".join(servers) or "none",
            ]
        )

    model_rows = []
    for model in sorted(models, key=lambda item: item["metadata"]["name"]):
        spec = model.get("spec", {})
        model_rows.append(
            [
                model["metadata"]["name"],
                spec.get("provider", "-"),
                spec.get("model", "-"),
                condition(model, "Accepted"),
            ]
        )

    kagent_ready, kagent_restarts, pod_rows = pod_summary(kagent_pods)
    ollama_ready, ollama_restarts, _ = pod_summary(ollama_pods)
    plane_ready, plane_restarts, _ = pod_summary(plane_pods)
    all_agents_ready = bool(agents) and all(
        condition(agent, "Ready") == "yes" and condition(agent, "Accepted") == "yes"
        for agent in agents
    )
    all_models_accepted = bool(models) and all(condition(model, "Accepted") == "yes" for model in models)
    runtime_ready = bool(kagent_pods) and kagent_ready == len(kagent_pods)
    if args.target == "kind":
        runtime_ready = runtime_ready and bool(ollama_pods) and ollama_ready == len(ollama_pods)
    overall_ready = all_agents_ready and all_models_accepted and runtime_ready

    print("Kaimahi status")
    print(f"  target:  {args.target}")
    print(f"  context: {args.context}")
    print(
        "  result:  "
        + (f"ready ({len(agents)} agents available)" if overall_ready else "attention required")
    )

    print("\nAgents")
    table(["NAME", "READY", "ACCEPTED", "MODEL CONFIG", "TOOL SERVER"], agent_rows)
    print("  Ready = the agent can serve requests; Accepted = kagent accepted its configuration.")

    print("\nModels")
    table(["CONFIG", "PROVIDER", "MODEL", "ACCEPTED"], model_rows)

    print("\nRuntime")
    print(f"  kagent:     {kagent_ready}/{len(kagent_pods)} pods ready, {kagent_restarts} restarts")
    if ollama_pods:
        print(f"  ollama:     {ollama_ready}/{len(ollama_pods)} pods ready, {ollama_restarts} restarts")
    else:
        print("  ollama:     not installed (expected for hosted-model targets)")
    if plane_pods:
        print(f"  governance: {plane_ready}/{len(plane_pods)} pods ready, {plane_restarts} restarts")
    else:
        print("  governance: not installed (run 'make plane' when you want budgets and audit)")

    print("\nRuntime pods")
    table(["NAME", "READY", "PHASE", "RESTARTS"], pod_rows)

    print("\nNext")
    if overall_ready:
        print("  make chat")
        tool_agent = next((row[0] for row in agent_rows if row[4] != "none"), None)
        if tool_agent:
            print(f"  make chat AGENT={tool_agent} TASK='What pods are running in the ollama namespace?'")
    else:
        print(f"  kubectl --context {args.context} -n kagent get agents,pods")
        print(f"  kubectl --context {args.context} -n kagent describe agent <name>")


if __name__ == "__main__":
    main()

File: internal/kmx/app/status.go
package app

import (
	"encoding/json"
	"fmt"
	"io"
	"sort"
	"strings"
)

// StatusOptions controls human or kubectl-native status output.
type StatusOptions struct {
	Output string
}

type objectList[T any] struct {
	Items []T `json:"items"`
}

type statusCondition struct {
	Type, Status string
}

type agentStatus struct {
	Metadata struct{ Name string } `json:"metadata"`
	Spec     struct {
		Declarative struct {
			ModelConfig string `json:"modelConfig"`
			Tools       []struct {
				MCPServer struct{ Name string } `json:"mcpServer"`
			} `json:"tools"`
		} `json:"declarative"`
	} `json:"spec"`
	Status struct{ Conditions []statusCondition } `json:"status"`
}

type modelStatus struct {
	Metadata struct{ Name string }                  `json:"metadata"`
	Spec     struct{ Provider, Model string }       `json:"spec"`
	Status   struct{ Conditions []statusCondition } `json:"status"`
}

type podStatus struct {
	Metadata struct{ Name string } `json:"metadata"`
	Status   struct {
		Phase             string
		Conditions        []statusCondition
		ContainerStatuses []struct {
			RestartCount int `json:"restartCount"`
		} `json:"containerStatuses"`
	} `json:"status"`
}

func condition(conditions []statusCondition, name string) string {
	for _, value := range conditions {
		if value.Type == name {
			if value.Status == "True" {
				return "yes"
			}
			return "no"
		}
	}
	return "-"
}

func table(out io.Writer, headers []string, rows [][]string) {
	widths := make([]int, len(headers))
	for i, header := range headers {
		widths[i] = len(header)
	}
	for _, row := range rows {
		for i, value := range row {
			if len(value) > widths[i] {
				widths[i] = len(value)
			}
		}
	}
	for rowIndex, row := range append([][]string{headers}, rows...) {
		fmt.Fprint(out, "  ")
		for i, value := range row {
			if i > 0 {
				fmt.Fprint(out, "  ")
			}
			fmt.Fprintf(out, "%-*s", widths[i], value)
		}
		if rowIndex < len(rows) {
			fmt.Fprintln(out)
		}
	}
	fmt.Fprintln(out)
}

func (a *App) statusJSON(namespace, resource string, optional bool, target any) error {
	raw, err := a.kubectlCapture("-n", namespace, "get", resource, "-o", "json")
	if err != nil {
		if optional && isNotFound(err) {
			return nil
		}
		return err
	}
	return json.Unmarshal([]byte(raw), target)
}

func podSummary(pods []podStatus) (ready, restarts int, rows [][]string) {
	sort.Slice(pods, func(i, j int) bool { return pods[i].Metadata.Name < pods[j].Metadata.Name })
	for _, pod := range pods {
		isReady := condition(pod.Status.Conditions, "Ready") == "yes"
		if isReady {
			ready++
		}
		podRestarts := 0
		for _, container := range pod.Status.ContainerStatuses {
			podRestarts += container.RestartCount
		}
		restarts += podRestarts
		rows = append(rows, []string{pod.Metadata.Name, map[bool]string{true: "yes", false: "no"}[isReady], pod.Status.Phase, fmt.Sprint(podRestarts)})
	}
	return
}

func statusReady(allAgents, allModels bool, kReady, kTotal, oReady, oTotal, pReady, pTotal int) bool {
	ready := allAgents && allModels && kTotal > 0 && kReady == kTotal
	if oTotal > 0 {
		ready = ready && oReady == oTotal
	}
	if pTotal > 0 {
		ready = ready && pReady == pTotal
	}
	return ready
}

// Status prints a grouped human view or kubectl-native JSON/YAML.
func (a *App) StatusWithOptions(opt StatusOptions) error {
	format := strings.ToLower(strings.TrimSpace(opt.Output))
	if format == "" || format == "table" {
		return a.statusTable()
	}
	if format != "json" && format != "yaml" {
		return fmt.Errorf("status output %q is not supported — use table, json, or yaml", opt.Output)
	}
	fmt.Fprintf(a.Err, "# context: %s (from %s)\n", a.Cfg.KubeContext, a.Cfg.ContextSource)
	return a.kubectlRun("-n", "kagent", "get", "agents,modelconfigs,pods", "-o", format)
}

func (a *App) Status() error { return a.StatusWithOptions(StatusOptions{}) }

func (a *App) statusTable() error {
	var agents objectList[agentStatus]
	var models objectList[modelStatus]
	var kagentPods, ollamaPods, planePods objectList[podStatus]
	if err := a.statusJSON("kagent", "agents", false, &agents); err != nil {
		return err
	}
	if err := a.statusJSON("kagent", "modelconfigs", false, &models); err != nil {
		return err
	}
	if err := a.statusJSON("kagent", "pods", false, &kagentPods); err != nil {
		return err
	}
	if err := a.statusJSON("ollama", "pods", true, &ollamaPods); err != nil {
		return err
	}
	if err := a.statusJSON("kaimahi", "pods", true, &planePods); err != nil {
		return err
	}

	agentRows := make([][]string, 0, len(agents.Items))
	allAgents := len(agents.Items) > 0
	for _, agent := range agents.Items {
		servers := make([]string, 0, len(agent.Spec.Declarative.Tools))
		for _, tool := range agent.Spec.Declarative.Tools {
			if tool.MCPServer.Name != "" {
				servers = append(servers, tool.MCPServer.Name)
			}
		}
		ready, accepted := condition(agent.Status.Conditions, "Ready"), condition(agent.Status.Conditions, "Accepted")
		allAgents = allAgents && ready == "yes" && accepted == "yes"
		agentRows = append(agentRows, []string{agent.Metadata.Name, ready, accepted, agent.Spec.Declarative.ModelConfig, valueOr(strings.Join(servers, ","), "none")})
	}
	sort.Slice(agentRows, func(i, j int) bool { return agentRows[i][0] < agentRows[j][0] })

	modelRows := make([][]string, 0, len(models.Items))
	allModels := len(models.Items) > 0
	for _, model := range models.Items {
		accepted := condition(model.Status.Conditions, "Accepted")
		allModels = allModels && accepted == "yes"
		modelRows = append(modelRows, []string{model.Metadata.Name, model.Spec.Provider, model.Spec.Model, accepted})
	}
	sort.Slice(modelRows, func(i, j int) bool { return modelRows[i][0] < modelRows[j][0] })

	kReady, kRestarts, podRows := podSummary(kagentPods.Items)
	oReady, oRestarts, _ := podSummary(ollamaPods.Items)
	pReady, pRestarts, _ := podSummary(planePods.Items)
	overall := statusReady(allAgents, allModels,
		kReady, len(kagentPods.Items), oReady, len(ollamaPods.Items), pReady, len(planePods.Items))

	fmt.Fprintln(a.Out, "Kaimahi status")
	fmt.Fprintf(a.Out, "  context: %s (from %s)\n", a.Cfg.KubeContext, a.Cfg.ContextSource)
	if overall {
		fmt.Fprintf(a.Out, "  result:  ready (%d agents available)\n", len(agents.Items))
	} else {
		fmt.Fprintln(a.Out, "  result:  attention required")
	}
	fmt.Fprintln(a.Out, "\nAgents")
	table(a.Out, []string{"NAME", "READY", "ACCEPTED", "MODEL CONFIG", "TOOL SERVER"}, agentRows)
	fmt.Fprintln(a.Out, "  Ready = can serve requests; Accepted = kagent accepted the configuration.")
	fmt.Fprintln(a.Out, "\nModels")
	table(a.Out, []string{"CONFIG", "PROVIDER", "MODEL", "ACCEPTED"}, modelRows)
	fmt.Fprintln(a.Out, "\nRuntime")
	fmt.Fprintf(a.Out, "  kagent:     %d/%d pods ready, %d restarts\n", kReady, len(kagentPods.Items), kRestarts)
	if len(ollamaPods.Items) > 0 {
		fmt.Fprintf(a.Out, "  ollama:     %d/%d pods ready, %d restarts\n", oReady, len(ollamaPods.Items), oRestarts)
	} else {
		fmt.Fprintln(a.Out, "  ollama:     not installed")
	}
	if len(planePods.Items) > 0 {
		fmt.Fprintf(a.Out, "  governance: %d/%d pods ready, %d restarts\n", pReady, len(planePods.Items), pRestarts)
	} else {
		fmt.Fprintln(a.Out, "  governance: not installed (run `kmx plane` for budgets and audit)")
	}
	fmt.Fprintln(a.Out, "\nRuntime pods")
	table(a.Out, []string{"NAME", "READY", "PHASE", "RESTARTS"}, podRows)
	fmt.Fprintln(a.Out, "\nNext")
	if overall {
		fmt.Fprintln(a.Out, "  kmx agent chat hello-world")
	} else {
		fmt.Fprintf(a.Out, "  kubectl --context %s -n kagent get agents,pods\n", a.Cfg.KubeContext)
	}
	return nil
}

func valueOr(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
